components/edit.js

```diff
@@ -0,0 +1,235 @@
+// node_modules
+import PropTypes from "prop-types"
+import React, { useContext, useState, useEffect } from "react"
+// components
+import Button from "./button"
+import EditJson, { EditLink, canEdit } from "./edit-func"
+import SessionContext from "./session-context"
+import { useAuthenticated } from './authentication'
+import { useRouter } from "next/router"
+// libs
+import Fetch from "../libs/sessionFetch"
+
+export const useEditor = (item, viewComponent) => {
+
+  const editing = (url) => { return url.endsWith("#!edit") }
+
+  const [edit, setEditing] = useState(false)
+
+  const router = useRouter()
+
+  useEffect(() => {
+    const isEdit = editing(document.URL)
+    // If the URL says edit but we aren't editing yet, set the state
+    if (isEdit && !edit) {
+      setEditing(true)
+    }
+
+    // If the URL has us not editing but we just were, set to false, and update props
+    if (!isEdit && edit) {
+      setEditing( false )
+      router.replace(router.asPath)
+    }
+  }, [edit, router])
+
+  const editpage = <EditPage item={item}/>
+  const editlink = <EditLink item={item}/>
+
+  const componentToShow = edit ? editpage :
+  <>
+    {viewComponent}
+    {editlink}
+  </>
+  return componentToShow
+}
+
+export const EditableItem = ({ item, children }) => {
+  return useEditor(item, children)
+}
```

Comment:
It’s so much nicer to use `EditableItem` on the pages now. Is there any value to having the `useEditor` custom hook though? That’s useful when you need the same React code shared across multiple components, but that’s not happening here. Is that likely to happen in the future, or can we just put all of `useEditor` within `EditableItem`?

Reply:
it seems kinda nice as an internal separation in the file? But I can copy it back into the EditableItem directly if you'd prefer.

Reply:
Is this a different case from other components that don’t use their own custom hook for their implementation?